dj/sql/dag.py

```diff
@@ -1,27 +1,35 @@
 """
 DAG related functions.
 """
+import collections
 from typing import List
 
-from dj.models.node import Node, NodeRevision
+from dj.models.node import Node
 from dj.utils import get_settings
 
 settings = get_settings()
 
 
 def get_dimensions(node: Node) -> List[str]:
     """
-    Return the available dimensions in a given node.
+    Return all available dimensions for a given node.
     """
     dimensions = []
-    for parent in node.current.parents:
-        for column in parent.current.columns:
-            dimensions.append(f"{parent.name}.{column.name}")
+    to_process = collections.deque([node, *node.current.parents])
+    processed = set()
 
-            if column.dimension:
+    while to_process:
+        current_node = to_process.popleft()
+        for column in current_node.current.columns:
+            if any(
+                attr.attribute_type.name == "dimension" for attr in column.attributes
+            ):
+                dimensions.append(f"{current_node.name}.{column.name}")
+            if column.dimension and column.dimension not in processed:
+                processed.add(column.dimension)
+                to_process.extend(column.dimension.current.parents)
```

Comment:
Based on our conversation [here](https://github.com/DataJunction/dj/issues/411), we may want to skip this part and stick to the closest parents only. Adding dimensions from all upstream parents may be hard/impossible.

Reply:
Yep, makes sense. And to support this correctly, we'll also need to add the dimension attribute "push-down" functionality -- as in, when someone links a column to a dimension node, DJ needs to push that link down to its children, where possible.

Reply:
Yes, as long as the column (with attributes) passed down is not altered any way. But its better to do this check during node building than during dimension discovery imo.

Reply:
Yeah I agree -- it shouldn't be during dimension discovery, but I was thinking it could happen at the dimensions linking stage. When a user links a column to a dimension node, that's when we push the dimension down to downstream nodes, where possible.

Reply:
@agorajek I removed this layer of adding the dimensions' upstreams (and simplified the function). I'll have a separate PR handle the attribute push-down, since it's closely linked to the build functionality.